gnome48Extensions: init

honnip opened the pull request:
- Introduce `--skip-fetch` option to update script. When this option is enabled, the script will not fetch extensions from EGO, but will still find name conflicts. 
- Set the user agent
- Fetch GNOME 48 extensions (but don't put them in `gnomeExtensions`)

With the `--skip-fetch` option, we can update `gnomeExtensions` for new GNOME version without having to download and update all extensions. 



## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

honnip commented:
It would be good to have this before GNOME 48 landing on the unstable channel, @jtojnar 

A follow-up commit is expected to include GNOME 48 extensions in the `gnomeExtensions` set, so should I target the `gnome` branch?

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -209,13 +206,6 @@ def process_extension(extension: Dict[str, Any]) -> Optional[Dict[str, Any]]:
 
-    for shell_version in shell_version_map.keys():
-        if pname in package_name_registry[shell_version]:
-            logging.warning(f"Package name '{pname}' for GNOME '{shell_version}' is colliding.")
```
Actually, reading a bit more, we might have two different collision checking mechanisms – intra-version here, and inter-(merged)-version below. But it does not seem like the former was being resolved in `collisions.json` in the prior state either.

honnip replied:
For the set of specific gnome version, it uses UUID for keys. So I think it's okay

jtojnar replied:
Makes sense.

jtojnar reviewed (commented):
Thanks for continuing handling this.

Looks mostly good.

> A follow-up commit is expected to include GNOME 48 extensions in the gnomeExtensions set, so should I target staging branch?

This can go to master for easier testing, with the follow up targetting staging-next.

honnip committed:
gnomeExtensions/update: Make collision calculation independent of fetching

Do not collect pname → uuid mapping during extension processing.
We will want to do it separately so that we can do it after changing
the set of merged versions without having to re-fetch everything.

Also remove the collision warning. It would only have warned about pname
collisions within a each `gnomeXYExtensions` attribute set but those
sets do not expose extensions by name, so the warning was pointless.

Since we are now iterating on `extensions.json`, we had to transpose the iterations.

honnip committed:
gnomeExtensions/update: Send User-Agent when scraping extensions

honnip committed:
gnomeExtensions/update: introduce `--skip-fetch` option

honnip committed:
gnomeExtensions/README: Split adding a new set and GNOME bumps

The former can be done independently on `master`.
The latter might be a longer process in a separate branch.

honnip committed:
gnome48Extensions: init

Updating `gnomeExtensions.gtk4-desktop-icons-ng-ding` broke the patch so it was rebased.
Additionally, there were some more programs that needed to be patched in:
- `update-desktop-database` and `gtk-update-icon-cache`: https://gitlab.com/smedius/desktop-icons-ng/commit/0a8e21b105967af28e0b1c085561e7e3bfa3cced
- `xdg-user-dirs`: https://gitlab.com/smedius/desktop-icons-ng/commit/e4c237ddf147f8e95789db4c93f40cceba7ef21d
Not adding the following, users will probably need to have them installed if they expect to work with those files:
- `7z` and `zipinfo`: https://gitlab.com/smedius/desktop-icons-ng/commit/b67c3ef94e64dd4ae2e0eac745e6f340df800335

Co-Authored-By: Jan Tojnar <jtojnar@gmail.com>

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -282,19 +270,21 @@ def scrape_extensions_index() -> List[Dict[str, Any]]:
+    return processed_extensions
+
+
+def serialize_extensions(processed_extensions: List[Dict[str, Any]]):
```
This should have `-> None` return value annotation for it to be type checked.

honnip commented:
I want to run the script again to be sure

honnip commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -295,54 +283,71 @@ def scrape_extensions_index() -> List[Dict[str, Any]]:
+    """
+    with open(updater_dir_path / "extensions.json", "r") as f:
+        extensions = json.load(f)
+    collisions: Dict[PackageName, Set[Uuid]] = {}
```
I picked `package_name_registry`

honnip commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -281,10 +272,7 @@ def scrape_extensions_index() -> List[Dict[str, Any]]:
-if __name__ == "__main__":
-    logging.basicConfig(level=logging.DEBUG)
-
+def fetch_extensions():
```
Good to know!

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -306,43 +299,57 @@ def scrape_extensions_index() -> List[Dict[str, Any]]:
+    """
+    Finds name collisions for the `major_versions`
+    """
+    with open(updater_dir_path / "extensions.json", "r") as f:
```
I think taking `extensions: list[Dict[str, Any]], versions_to_merge: list[str]` as an argument and doing the IO and passing globals in main might be cleaner and also allow testing the function in tests/python console.

jtojnar replied:
In the future, this will allow us to add something like:

```python
class FindCollisions(unittest.TestCase):
    def no_collision(self):
        self.assertEqual(
            {},
            find_collisions(
                extensions=[
                    {
                        "name": "removable-drive-menu",
                        "uuid": "drive-menu@gnome-shell-extensions.gcampax.github.com",
                        "shell_version_map": {"38": {}, "40": {}, "41": {}, "42": {}, "43": {}, "44": {}, "45": {}, "46": {}, "47": {}},
                    },
                ],
                versions_to_merge=["46", "47", "48"],
            ),
        )

    def collision(self):
        self.assertEqual(
            {
                "system-monitor": "system-monitor": [
                    "System_Monitor@bghome.gmail.com",
                    "system-monitor@gnome-shell-extensions.gcampax.github.com",
                ],
            },
            find_collisions(
                extensions=[
                    {
                        "name": "system-monitor",
                        "uuid": "system-monitor@gnome-shell-extensions.gcampax.github.com",
                        "shell_version_map": {"46": {}, "47": {}, "48": {}},
                    },
                    {
                        "name": "system-monitor",
                        "uuid": "System_Monitor@bghome.gmail.com",
                        "shell_version_map": {"38": {}, "40": {}, "41": {}, "42": {}, "43": {}, "44": {}, "45": {}, "46": {}, "47": {}},
                    },
                ],
                versions_to_merge=["46", "47", "48"],
            ),
        )

```

Then the tests can be run using `python3 -m unittest pkgs/desktops/gnome/extensions/update-extensions.py`

honnip replied:
Looks really good to me. I'll make a PR about this in a few days

honnip replied:
Done in [`b642b9d` (#403591)](https://github.com/NixOS/nixpkgs/pull/403591/commits/b642b9d307f928e60dce0e57284011235ab04710), with type name changes (https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections)

honnip committed:
gnomeExtensions: Add unit tests and development instructions

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -345,14 +348,63 @@ def find_collisions() -> Dict[PackageName, List[Uuid]]:
+        )
+
+    def test_collision(self):
+            self.assertEqual(
+                {
+                    "bar": [
+                        "bar_42_to_45@chulsoo.example",
+                        "bar_44_to_47@younghee.example",
+                    ],
+                },
+                find_collisions(self.extensions, ["45", "46", "47"]),
+            )
```
The indentation is off here.

jtojnar commented on pkgs/desktops/gnome/extensions/README.md:
```diff
@@ -1,26 +1,31 @@
+#### For GNOME updates,
+
+1. Follow the [To package the extensions for new GNOME version](#to-package-the-extensions-for-new-gnome-version) section if required.
+2. Update `major_versions` variable in `./update-extensions.py`.
```
This is now versions_to_merge

jtojnar committed:
gnomeExtensions/update: Format with black

We have not formatted it for a while, let’s catch up.

honnip committed:
gnomeExtensions/update: Modernize type annotations

Since Python 3.9, standard collections can be used with generics in type annotations:
https://docs.python.org/3.9/whatsnew/3.9.html#type-hinting-generics-in-standard-collections

Since Python 3.10, `Optional` can be written using type union with `None`:
https://docs.python.org/3.10/whatsnew/3.10.html#pep-604-new-type-union-operator

honnip committed:
gnomeExtensions/update: Fix Incompatible types in assignment error

Type checking with mypy would complain:

    error: Incompatible types in assignment (expression has type "str", variable has type "dict[str, Any]")  [assignment]

honnip committed:
gnomeExtensions/update: Split into functions

This will make it easier to reason about the code and test it.

There are no functional or other changes.

honnip committed:
gnomeExtensions/update: Enumerate GNOME versions to merge

Previously, we automatically used the last three versions listed
in `supported_versions` to produce the `collisions.json` file.
In the definition of the `gnomeExtensions` attribute set,
the versions are, however, listed explicitly.

Let’s list them explicitly in the updater as well.

jtojnar committed:
gnomeExtensions/update: Simplify collision calculation

- Use meaningful variable names in list comprehensions.
- Remove redundant comments.
- Use `setdefault` instead of conditional assignment.
- Move the filtering into the returned list comprehension.
- Rename intermediary variable since we are now filtering in the return statement.

honnip committed:
gnomeExtensions/update: Sort colliding UUIDs

Unlike `dict`, whose items are guaranteed to follow the insertion order since Python 3.7:
https://docs.python.org/3.13/library/stdtypes.html#mapping-types-dict
Set has no such guarantees:
https://docs.python.org/3.13/library/stdtypes.html#set-types-set-frozenset

Let’s sort the UUIDs to ensure consistent order.

jtojnar commented:
I took the liberty to force push https://github.com/NixOS/nixpkgs/compare/b642b9d307f928e60dce0e57284011235ab04710..752f12cec68a73faf6e8913fc319971e4e4a65bd with the fixes.

Additionally, I have:

- Split the commits even further as I reviewed them. Included reasoning behind the changes according to my best understanding.
- Formatted the code with black, which is another validation we have not done for a while. Also added it to the comment at the top.
- Replaced `Optional[T]` with more modern `T | None`.
- Moved the body of entry point into `main` function.
- Removed redundant `list` in `sorted(list(…))`.
- Renamed `k` and `v` in list comprehensions to a more meaningful name.
- Added return types to tests.

jtojnar reviewed (approved):
I tried updating following the new procedure, appears to work.

jtojnar committed:
nixosTests.gnome-extension: Wait for activation

This fixes the following failure:

Test "Enable extension 'dash-to-panel@jderose9.github.com'" failed with error: "  State: ACTIVATING instead of ACTIVE"

I can reproduce it even before the parent commit.

jtojnar commented:
Fixed a patch for ding, and applied a workaround for a dash-to-panel test failing for me.

honnip commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 403591`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 340 packages built:</summary>
  <ul>
    <li>gnomeExtensions.activate-window-by-title</li>
    <li>gnomeExtensions.activate_gnome</li>
    <li>gnomeExtensions.add-custom-text-to-workspace-indicators</li>
    <li>gnomeExtensions.add-username-to-top-panel</li>
    <li>gnomeExtensions.advanced-weather-companion</li>
    <li>gnomeExtensions.adw-gtk3-colorizer</li>
    <li>gnomeExtensions.all-windows</li>
    <li>gnomeExtensions.all-windows-saverestore-window-positions</li>
    <li>gnomeExtensions.alt-tab-current-monitor</li>
    <li>gnomeExtensions.alttab-launcher</li>
    <li>gnomeExtensions.alttab-scroll-workaround</li>
    <li>gnomeExtensions.always-indicator</li>
    <li>gnomeExtensions.always-show-workspace-thumbnails</li>
    <li>gnomeExtensions.app-hider</li>
    <li>gnomeExtensions.appindicator</li>
    <li>gnomeExtensions.application-tabs</li>
    <li>gnomeExtensions.applications-overview-tooltip</li>
    <li>gnomeExtensions.astra-monitor</li>
    <li>gnomeExtensions.auto-accent-colour</li>
    <li>gnomeExtensions.auto-adwaita-colors</li>
    <li>gnomeExtensions.auto-power-profile</li>
    <li>gnomeExtensions.auto-select-headset</li>
    <li>gnomeExtensions.autohide-battery</li>
    <li>gnomeExtensions.autohide-battery-percentage</li>
    <li>gnomeExtensions.autohide-volume</li>
    <li>gnomeExtensions.bangs-search</li>
    <li>gnomeExtensions.battery-consumption-watt-meter</li>
    <li>gnomeExtensions.battery-indicator-icon</li>
    <li>gnomeExtensions.battery-time</li>
    <li>gnomeExtensions.battery-time-2</li>
    <li>gnomeExtensions.battery-time-3</li>
    <li>gnomeExtensions.battery-time-percentage-compact</li>
    <li>gnomeExtensions.battery-usage-wattmeter</li>
    <li>gnomeExtensions.better-ibus</li>
    <li>gnomeExtensions.bilingual-app-search</li>
    <li>gnomeExtensions.bing-wallpaper</li>
    <li>gnomeExtensions.bing-wallpaper-changer</li>
    <li>gnomeExtensions.binu</li>
    <li>gnomeExtensions.bitahub</li>
    <li>gnomeExtensions.bitcoin-markets</li>
    <li>gnomeExtensions.bluetooth-battery-meter</li>
    <li>gnomeExtensions.bluetooth-file-sender</li>
    <li>gnomeExtensions.bluetooth-quick-connect</li>
    <li>gnomeExtensions.blurt</li>
    <li>gnomeExtensions.boost-volume</li>
    <li>gnomeExtensions.brightness-control-using-ddcutil</li>
    <li>gnomeExtensions.bring-out-submenu-of-power-offlogout-button</li>
    <li>gnomeExtensions.bubblemail</li>
    <li>gnomeExtensions.caffeine</li>
    <li>gnomeExtensions.calc</li>
    <li>gnomeExtensions.caplang-indicator</li>
    <li>gnomeExtensions.click-to-close-overview</li>
    <li>gnomeExtensions.clipboard-indicator</li>
    <li>gnomeExtensions.clipqr</li>
    <li>gnomeExtensions.cmud</li>
    <li>gnomeExtensions.commands-store</li>
    <li>gnomeExtensions.compact-top-bar</li>
    <li>gnomeExtensions.containers</li>
    <li>gnomeExtensions.control-monitor-brightness-and-volume-with-ddcutil</li>
    <li>gnomeExtensions.copier</li>
    <li>gnomeExtensions.crazy-internet-speed-meter</li>
    <li>gnomeExtensions.crypto-tracker</li>
    <li>gnomeExtensions.cryptogoldbitcoin-rate</li>
    <li>gnomeExtensions.currency-tracker</li>
    <li>gnomeExtensions.current-monitor-windowapp-switcher</li>
    <li>gnomeExtensions.custom-command-list</li>
    <li>gnomeExtensions.custom-hot-corners-extended</li>
    <li>gnomeExtensions.custom-menu</li>
    <li>gnomeExtensions.custom-reboot</li>
    <li>gnomeExtensions.customize-clock-on-lock-screen</li>
    <li>gnomeExtensions.dash-in-panel</li>
    <li>gnomeExtensions.dash2dock-lite</li>
    <li>gnomeExtensions.date-menu-formatter</li>
    <li>gnomeExtensions.ddterm</li>
    <li>gnomeExtensions.debian-linux-updates-indicator</li>
    <li>gnomeExtensions.default-workspace</li>
    <li>gnomeExtensions.dejaview</li>
    <li>gnomeExtensions.desktop-icons-ng-ding</li>
    <li>gnomeExtensions.dev-container-manager</li>
    <li>gnomeExtensions.dexcom-cgm-monitor</li>
    <li>gnomeExtensions.dfx-currency-converter</li>
    <li>gnomeExtensions.dim-background-windows</li>
    <li>gnomeExtensions.dim-completed-calendar-events</li>
    <li>gnomeExtensions.disable-3-finger-gestures</li>
    <li>gnomeExtensions.disable-workspace-switcher-overlay</li>
    <li>gnomeExtensions.discrete-brightness</li>
    <li>gnomeExtensions.displays-adjustments</li>
    <li>gnomeExtensions.dnd-on-time</li>
    <li>gnomeExtensions.do-not-disturb-while-screen-sharing-or-recording</li>
    <li>gnomeExtensions.downfall</li>
    <li>gnomeExtensions.dragntile</li>
    <li>gnomeExtensions.dual-shock-4-battery-percentage</li>
    <li>gnomeExtensions.dynamic-panel</li>
    <li>gnomeExtensions.easy-docker-containers</li>
    <li>gnomeExtensions.easyeffects-preset-selector</li>
    <li>gnomeExtensions.emacs-search-provider</li>
    <li>gnomeExtensions.emoji-copy</li>
    <li>gnomeExtensions.escape-overview</li>
    <li>gnomeExtensions.exchange-electricity-price-indicator</li>
    <li>gnomeExtensions.executor</li>
    <li>gnomeExtensions.favorites-to-applications-grid</li>
    <li>gnomeExtensions.firefox-profiles</li>
    <li>gnomeExtensions.fix-remote-window-icons</li>
    <li>gnomeExtensions.fixed-ime-list</li>
    <li>gnomeExtensions.fly-pie</li>
    <li>gnomeExtensions.fnlock-switch-thinkpad-compact-usb-keyboard</li>
    <li>gnomeExtensions.focus-follows-workspace</li>
    <li>gnomeExtensions.focused-window-d-bus</li>
    <li>gnomeExtensions.foresight</li>
    <li>gnomeExtensions.framework-fan-control</li>
    <li>gnomeExtensions.frequency-boost-switch</li>
    <li>gnomeExtensions.frippery-applications-menu</li>
    <li>gnomeExtensions.fullscreen-avoider</li>
    <li>gnomeExtensions.fullscreen-button</li>
    <li>gnomeExtensions.fullscreen-to-empty-workspace-2</li>
    <li>gnomeExtensions.fuzzy-app-search</li>
    <li>gnomeExtensions.gamebar-overlay</li>
    <li>gnomeExtensions.gamemode-shell-extension</li>
    <li>gnomeExtensions.gbinaryclock</li>
    <li>gnomeExtensions.gmeet</li>
    <li>gnomeExtensions.gnome-40-ui-improvements</li>
    <li>gnomeExtensions.gnordvpn-local</li>
    <li>gnomeExtensions.go-to-last-workspace</li>
    <li>gnomeExtensions.gpu-profile-selector</li>
    <li>gnomeExtensions.gpu-supergfxctl-switch</li>
    <li>gnomeExtensions.gradient-top-bar</li>
    <li>gnomeExtensions.grayscale-windows</li>
    <li>gnomeExtensions.gsnap</li>
    <li>gnomeExtensions.gtile</li>
    <li>gnomeExtensions.gtk4-desktop-icons-ng-ding</li>
    <li>gnomeExtensions.guake-ssh</li>
    <li>gnomeExtensions.happy-appy-hotkey</li>
    <li>gnomeExtensions.headsetcontrol</li>
    <li>gnomeExtensions.herkul</li>
    <li>gnomeExtensions.hibernate-status-button</li>
    <li>gnomeExtensions.hide-keyboard-layout</li>
    <li>gnomeExtensions.hide-minimized</li>
    <li>gnomeExtensions.hide-screen-sharing</li>
    <li>gnomeExtensions.hide-the-dock-in-overview</li>
    <li>gnomeExtensions.hide-top-bar</li>
    <li>gnomeExtensions.hide-volume-indicator</li>
    <li>gnomeExtensions.highlight-focus</li>
    <li>gnomeExtensions.historymanager-prefix-search</li>
    <li>gnomeExtensions.home-assistant-extension</li>
    <li>gnomeExtensions.hostname-on-panel</li>
    <li>gnomeExtensions.huawei-wmi-controls</li>
    <li>gnomeExtensions.ibus-font-setting</li>
    <li>gnomeExtensions.ideapad-controls</li>
    <li>gnomeExtensions.indicator</li>
    <li>gnomeExtensions.inhibition-indicator</li>
    <li>gnomeExtensions.input-source-d-bus-interface</li>
    <li>gnomeExtensions.internet-speed-meter</li>
    <li>gnomeExtensions.ip-address-extension</li>
    <li>gnomeExtensions.ip-indicator</li>
    <li>gnomeExtensions.iscsi-quick-connect</li>
    <li>gnomeExtensions.iso8601-ish-clock</li>
    <li>gnomeExtensions.kali-track</li>
    <li>gnomeExtensions.keep-awake</li>
    <li>gnomeExtensions.keep-pinned-apps-in-appgrid</li>
    <li>gnomeExtensions.kerberos-login</li>
    <li>gnomeExtensions.kernel-indicator</li>
    <li>gnomeExtensions.keyboard-modifiers-status</li>
    <li>gnomeExtensions.keyboard-reset</li>
    <li>gnomeExtensions.keyring-autolock</li>
    <li>gnomeExtensions.lan-ip-address</li>
    <li>gnomeExtensions.launcher</li>
    <li>gnomeExtensions.lilypad</li>
    <li>gnomeExtensions.lineup</li>
    <li>gnomeExtensions.lockscreen-extension</li>
    <li>gnomeExtensions.logo-activities</li>
    <li>gnomeExtensions.lomotion</li>
    <li>gnomeExtensions.looking-glass-button</li>
    <li>gnomeExtensions.luna-moon-phase-indicator</li>
    <li>gnomeExtensions.macpeek</li>
    <li>gnomeExtensions.media-progress</li>
    <li>gnomeExtensions.memento-mori</li>
    <li>gnomeExtensions.messaging-menu</li>
    <li>gnomeExtensions.middle-click-to-close-in-overview</li>
    <li>gnomeExtensions.mock-tray</li>
    <li>gnomeExtensions.move-clock</li>
    <li>gnomeExtensions.move-to-next-screen</li>
    <li>gnomeExtensions.moveclock</li>
    <li>gnomeExtensions.mullvad-indicator</li>
    <li>gnomeExtensions.multicore-system-monitor</li>
    <li>gnomeExtensions.mute-spotify-ads</li>
    <li>gnomeExtensions.muteunmute</li>
    <li>gnomeExtensions.nasa-apod</li>
    <li>gnomeExtensions.nepali-calendar</li>
    <li>gnomeExtensions.nepali-calendar-2</li>
    <li>gnomeExtensions.net-speed-indicator</li>
    <li>gnomeExtensions.net-speed-simplified</li>
    <li>gnomeExtensions.new-mail-indicator</li>
    <li>gnomeExtensions.new-workspace-shortcut</li>
    <li>gnomeExtensions.nextdns-indicator</li>
    <li>gnomeExtensions.nightscout</li>
    <li>gnomeExtensions.noiseclapper</li>
    <li>gnomeExtensions.nothing-to-say</li>
    <li>gnomeExtensions.notification-banner-position</li>
    <li>gnomeExtensions.notification-banner-reloaded</li>
    <li>gnomeExtensions.notification-counter</li>
    <li>gnomeExtensions.notifications-alert-on-user-menu</li>
    <li>gnomeExtensions.ollama-indicator</li>
    <li>gnomeExtensions.one-thing</li>
    <li>gnomeExtensions.one-window-wonderland</li>
    <li>gnomeExtensions.openmeteoweather</li>
    <li>gnomeExtensions.osd-volume-number</li>
    <li>gnomeExtensions.ostree-status</li>
    <li>gnomeExtensions.otp-keys</li>
    <li>gnomeExtensions.overview-flick</li>
    <li>gnomeExtensions.oxford-3000-word-trainer-tr-en</li>
    <li>gnomeExtensions.panel-favorites</li>
    <li>gnomeExtensions.panel-workspace-scroll</li>
    <li>gnomeExtensions.panel-world-clock-lite</li>
    <li>gnomeExtensions.panelhideswindow</li>
    <li>gnomeExtensions.paperwm</li>
    <li>gnomeExtensions.peek-top-bar-on-fullscreen</li>
    <li>gnomeExtensions.persian-calendar</li>
    <li>gnomeExtensions.persian-calendar-2</li>
    <li>gnomeExtensions.phi-pi-hole-indicator</li>
    <li>gnomeExtensions.picture-of-the-day</li>
    <li>gnomeExtensions.pin-it</li>
    <li>gnomeExtensions.ping</li>
    <li>gnomeExtensions.pip-on-top</li>
    <li>gnomeExtensions.pipewire-airplay-toggle</li>
    <li>gnomeExtensions.pipewire-settings</li>
    <li>gnomeExtensions.pointer-tracker</li>
    <li>gnomeExtensions.port-forwarding-extension</li>
    <li>gnomeExtensions.power-tracker</li>
    <li>gnomeExtensions.preedit-highlight-for-japanese-ime</li>
    <li>gnomeExtensions.presentation-mode-panel</li>
    <li>gnomeExtensions.primary-input-on-lockscreen</li>
    <li>gnomeExtensions.printers</li>
    <li>gnomeExtensions.quarter-windows</li>
    <li>gnomeExtensions.quick-lofi</li>
    <li>gnomeExtensions.quick-settings-audio-panel</li>
    <li>gnomeExtensions.quick-touchpad-toggle</li>
    <li>gnomeExtensions.quick-web-search</li>
    <li>gnomeExtensions.quicksettings-indicator-visibility-tool</li>
    <li>gnomeExtensions.quran-player</li>
    <li>gnomeExtensions.radio-kayra</li>
    <li>gnomeExtensions.razer-puppy</li>
    <li>gnomeExtensions.rdesktop-launcher</li>
    <li>gnomeExtensions.reboottouefi</li>
    <li>gnomeExtensions.rectangle</li>
    <li>gnomeExtensions.remmina-search-provider</li>
    <li>gnomeExtensions.reorder-workspaces</li>
    <li>gnomeExtensions.resolution-and-refresh-rate-in-quick-settings</li>
    <li>gnomeExtensions.restart-to</li>
    <li>gnomeExtensions.rounded-corners</li>
    <li>gnomeExtensions.run-android-emulators</li>
    <li>gnomeExtensions.runcat</li>
    <li>gnomeExtensions.screen-rotate</li>
    <li>gnomeExtensions.screenshort-cut</li>
    <li>gnomeExtensions.server-status-indicator</li>
    <li>gnomeExtensions.settingscenter</li>
    <li>gnomeExtensions.show-desktop-button</li>
    <li>gnomeExtensions.show-favorite-apps</li>
    <li>gnomeExtensions.shutdown-dialogue</li>
    <li>gnomeExtensions.shutdowntimer</li>
    <li>gnomeExtensions.simple-message</li>
    <li>gnomeExtensions.simple-timer</li>
    <li>gnomeExtensions.slinger</li>
    <li>gnomeExtensions.smart-home</li>
    <li>gnomeExtensions.smile-complementary-extension</li>
    <li>gnomeExtensions.soft-brightness-plus</li>
    <li>gnomeExtensions.solaar-extension</li>
    <li>gnomeExtensions.sound-loopback</li>
    <li>gnomeExtensions.sound-percentage</li>
    <li>gnomeExtensions.space-bar</li>
    <li>gnomeExtensions.speedinator</li>
    <li>gnomeExtensions.spotify-controls</li>
    <li>gnomeExtensions.ssh-profile-list</li>
    <li>gnomeExtensions.statistig</li>
    <li>gnomeExtensions.status-area-horizontal-spacing</li>
    <li>gnomeExtensions.stopwatch</li>
    <li>gnomeExtensions.streamcontroller-integration</li>
    <li>gnomeExtensions.switch-two-layouts</li>
    <li>gnomeExtensions.switch-x11-wayland-default-session</li>
    <li>gnomeExtensions.syncthing-indicator</li>
    <li>gnomeExtensions.syncthing-toggle</li>
    <li>gnomeExtensions.system-software-info</li>
    <li>gnomeExtensions.tailscale-qs</li>
    <li>gnomeExtensions.task-up-ultralite</li>
    <li>gnomeExtensions.teatimer</li>
    <li>gnomeExtensions.tempomate</li>
    <li>gnomeExtensions.text-clock</li>
    <li>gnomeExtensions.thinkpad-red-led</li>
    <li>gnomeExtensions.thinkpad-thermal</li>
    <li>gnomeExtensions.tiling-shell</li>
    <li>gnomeExtensions.time-from-start-uptime</li>
    <li>gnomeExtensions.todo-list</li>
    <li>gnomeExtensions.toggler</li>
    <li>gnomeExtensions.top-panel-workspace-scroll</li>
    <li>gnomeExtensions.topiconsfix</li>
    <li>gnomeExtensions.totp</li>
    <li>gnomeExtensions.touchpad-gesture-customization</li>
    <li>gnomeExtensions.touchpad-switcher</li>
    <li>gnomeExtensions.tracker</li>
    <li>gnomeExtensions.trainfi</li>
    <li>gnomeExtensions.transcodeappsearch</li>
    <li>gnomeExtensions.transmission-daemon-indicator-ng</li>
    <li>gnomeExtensions.transparent-top-bar-adjustable-transparency</li>
    <li>gnomeExtensions.transparent-window-moving</li>
    <li>gnomeExtensions.trimmer</li>
    <li>gnomeExtensions.unmess</li>
    <li>gnomeExtensions.upower-battery</li>
    <li>gnomeExtensions.usd-to-rub-exchange-rate</li>
    <li>gnomeExtensions.usd-toman</li>
    <li>gnomeExtensions.useless-gaps</li>
    <li>gnomeExtensions.user-avatar-in-quick-settings</li>
    <li>gnomeExtensions.username-in-top-panel-right-side</li>
    <li>gnomeExtensions.vertical-workspaces</li>
    <li>gnomeExtensions.vim-alt-tab</li>
    <li>gnomeExtensions.vkeybind</li>
    <li>gnomeExtensions.void</li>
    <li>gnomeExtensions.voluble</li>
    <li>gnomeExtensions.volume-scroller-2</li>
    <li>gnomeExtensions.vscode-search-provider</li>
    <li>gnomeExtensions.vscode-workspaces-gnome</li>
    <li>gnomeExtensions.vuemeter-system</li>
    <li>gnomeExtensions.warp-toggle</li>
    <li>gnomeExtensions.weather-or-not</li>
    <li>gnomeExtensions.week-start-modifier</li>
    <li>gnomeExtensions.what-watch</li>
    <li>gnomeExtensions.wifi-qrcode</li>
    <li>gnomeExtensions.wikiart-wallpaper</li>
    <li>gnomeExtensions.window-calls</li>
    <li>gnomeExtensions.window-centering</li>
    <li>gnomeExtensions.window-gestures</li>
    <li>gnomeExtensions.window-title-is-back</li>
    <li>gnomeExtensions.windowswitcher</li>
    <li>gnomeExtensions.wintile-windows-10-window-tiling-for-gnome</li>
    <li>gnomeExtensions.wireguard-vpn-extension</li>
    <li>gnomeExtensions.workspace-buttons-with-app-icons</li>
    <li>gnomeExtensions.workspaces-indicator-by-open-apps</li>
    <li>gnomeExtensions.x11-gestures</li>
    <li>gnomeExtensions.yakuake</li>
    <li>gnomeExtensions.yks-timer</li>
    <li>gnomeExtensions.zen</li>
    <li>gnomeExtensions.zim-quick-launch</li>
  </ul>
</details>

honnip commented:
These are removed on EGO:
- gnome-shell-extension-control-blur-effect-on-lock-screen (†26)
- gnome-shell-extension-end-session-timer (†13)
- gnome-shell-extension-replace-activities-text (†22)

honnip commented:
Everything looks better than before. Thank you.

I was unable to build the test because of the many rebuilds.

honnip commented:
I noticed that ding extensions is not working, though it is enabled successfully.

```
 5월 09 16:43:22 acrux .gnome-shell-wr[65960]: Adw-DING enabled.
 5월 09 16:43:22 acrux .gnome-shell-wr[65960]: com.desktop.dingextension DBus Name Acquired
 5월 09 16:43:22 acrux .gnome-shell-wr[65960]: Adw-DING: (gjs:96012): GLib-GObject-CRITICAL **: 16:43:22.652: cannot register existing type 'GCancellable'
 5월 09 16:43:22 acrux .gnome-shell-wr[65960]: Adw-DING: (gjs:96012): GLib-GObject-CRITICAL **: 16:43:22.652: cannot add private field to invalid (non-instantiatable) type '<invalid>'
 5월 09 16:43:22 acrux .gnome-shell-wr[65960]: Adw-DING: (gjs:96012): GLib-CRITICAL **: 16:43:22.652: g_once_init_leave_pointer: assertion 'result != 0' failed
```

I have no idea about this

jtojnar commented:
I built a GNOME VM with `gnomeExtensions.gtk4-desktop-icons-ng-ding` installed and do not see the assertions above, the icons appear on the desktop after enabling the extension.

However, there were some more missing programs, I patched them.

Additionally, some component crashes due to libsoup conflict but this also does not appear to have an effect on the icons:

<details>
<summary>Details</summary>


```
May 09 11:40:31 nixos .gnome-shell-wr[1676]: Launching Adw-DING process
May 09 11:40:31 nixos .gnome-shell-wr[1676]: Adw-DING enabled.
May 09 11:40:31 nixos .gnome-shell-wr[1676]: com.desktop.dingextension DBus Name Acquired
May 09 11:40:32 nixos .gnome-shell-wr[1676]: Adw-DING: (gjs:2268): Gjs-Console-CRITICAL **: 11:40:32.145: Install Poppler for proper fallback pdf thumbnailing 
May 09 11:40:32 nixos .gnome-shell-wr[1676]: Adw-DING:  Error: Requiring Poppler, version none: Typelib file for namespace 'Poppler' (any version) not found
May 09 11:40:32 nixos .gnome-shell-wr[1676]: Adw-DING: Gjs-Console-Message: 11:40:32.280: Local install detected, updating icon cache...
May 09 11:40:32 nixos dbus-daemon[1580]: [session uid=1000 pid=1580] Activating service name='org.gnome.Nautilus' requested by ':1.84' (uid=1000 pid=2268 comm="/nix/store/xb548l4yc5sanaqpv3zm3j815dy4a36d-gjs-1." label="kernel")
May 09 11:40:32 nixos .nautilus-wrapp[2281]: Connecting to org.freedesktop.Tracker3.Miner.Files
May 09 11:40:32 nixos dbus-daemon[1580]: [session uid=1000 pid=1580] Successfully activated service 'org.gnome.Nautilus'
May 09 11:40:32 nixos dbus-daemon[1580]: [session uid=1000 pid=1580] Activating service name='org.gnome.NautilusPreviewer' requested by ':1.84' (uid=1000 pid=2268 comm="/nix/store/xb548l4yc5sanaqpv3zm3j815dy4a36d-gjs-1." label="kernel")
May 09 11:40:32 nixos dbus-daemon[1580]: [session uid=1000 pid=1580] Activating service name='org.gnome.ArchiveManager1' requested by ':1.84' (uid=1000 pid=2268 comm="/nix/store/xb548l4yc5sanaqpv3zm3j815dy4a36d-gjs-1." label="kernel")
May 09 11:40:32 nixos .gnome-shell-wr[1676]: Adw-DING: Gjs-Console-Message: 11:40:32.626: System supports accent colors
May 09 11:40:32 nixos .gnome-shell-wr[1676]: Adw-DING: (gjs:2268): Gjs-Console-CRITICAL **: 11:40:32.816: Desktop directory does not exist
May 09 11:40:32 nixos dbus-daemon[1580]: [session uid=1000 pid=1580] Activating via systemd: service name='org.freedesktop.Tracker3.Miner.Files' unit='localsearch-3.service' requested by ':1.85' (uid=1000 pid=2281 comm="/nix/store/73da2nfarhd0fxsvf85jrslx5fz1h5pm-nautil" label="kernel")
May 09 11:40:32 nixos dbus-daemon[1580]: [session uid=1000 pid=1580] Successfully activated service 'org.gnome.ArchiveManager1'
May 09 11:40:32 nixos systemd[1571]: Created slice User Background Tasks Slice.
░░ Subject: A start job for unit UNIT has finished successfully
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ A start job for unit UNIT has finished successfully.
░░ 
░░ The job identifier is 434.
May 09 11:40:32 nixos systemd[1571]: Starting Tracker file system data miner...
░░ Subject: A start job for unit UNIT has begun execution
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ A start job for unit UNIT has begun execution.
░░ 
░░ The job identifier is 421.
May 09 11:40:33 nixos .gnome-shell-wr[1676]: Adw-DING: Gjs-Console-Message: 11:40:33.712: Updated: com.desktop.ding.desktop
May 09 11:40:33 nixos .gnome-shell-wr[1676]: Adw-DING: Gjs-Console-Message: 11:40:33.798: Updated desktop database
May 09 11:40:33 nixos .gnome-shell-wr[1676]: Making Wayland window type Desktop
May 09 11:40:34 nixos gst-plugin-scan[2327]: libsoup3 symbols detected. Using libsoup2 and libsoup3 in the same process is not supported.
May 09 11:40:34 nixos kernel: traps: gst-plugin-scan[2327] trap int3 ip:7f453cb01a0f sp:7ffc14abdb50 error:0 in libglib-2.0.so.0.8200.5[6ba0f,7f453cab9000+a3000]
May 09 11:40:34 nixos systemd-coredump[2335]: Process 2327 (gst-plugin-scan) of user 1000 terminated abnormally with signal 5/TRAP, processing...
```

</details>
